src/wasm-type.h

```diff
@@ -270,70 +271,24 @@ class Type {
 
   std::string toString() const;
 
-  struct Iterator {
-    // Iterator traits
-    using iterator_category = std::random_access_iterator_tag;
+  size_t size() const;
+
+  struct Iterator : ParentIndexIterator<const Type*, Iterator> {
     using value_type = Type;
-    using difference_type = std::ptrdiff_t;
     using pointer = const Type*;
     using reference = const Type&;
-
-    const Type* parent;
-    size_t index;
-    Iterator(const Type* parent, size_t index) : parent(parent), index(index) {}
-    bool operator==(const Iterator& other) const {
-      return index == other.index && parent == other.parent;
-    }
-    bool operator!=(const Iterator& other) const { return !(*this == other); }
-    Iterator& operator++() {
-      ++index;
-      return *this;
-    }
-    Iterator& operator--() {
-      --index;
-      return *this;
-    }
-    Iterator operator++(int) {
-      auto it = *this;
-      index++;
-      return it;
-    }
-    Iterator operator--(int) {
-      auto it = *this;
-      index--;
-      return it;
-    }
-    Iterator& operator+=(difference_type off) {
-      index += off;
-      return *this;
-    }
-    Iterator operator+(difference_type off) const {
-      return Iterator(*this) += off;
-    }
-    Iterator& operator-=(difference_type off) {
-      index -= off;
-      return *this;
-    }
-    Iterator operator-(difference_type off) const {
-      return Iterator(*this) -= off;
-    }
-    difference_type operator-(const Iterator& other) const {
-      assert(parent == other.parent);
-      return index - other.index;
-    }
-    const value_type& operator*() const;
+    reference operator*() const;
   };
 
-  Iterator begin() const { return Iterator(this, 0); }
-  Iterator end() const;
+  Iterator begin() const { return Iterator{{this, 0}}; }
+  Iterator end() const { return Iterator{{this, size()}}; }
   std::reverse_iterator<Iterator> rbegin() const {
     return std::make_reverse_iterator(end());
   }
   std::reverse_iterator<Iterator> rend() const {
     return std::make_reverse_iterator(begin());
   }
-  size_t size() const { return end() - begin(); }
-  const Type& operator[](size_t i) const;
+  const Type& operator[](size_t i) const { return *Iterator{{this, i}}; }
```

Comment:
This change seems like it might be slower now, perhaps worth benchmarking?

Reply:
My unscientific benchmark running `-O2` on a poppler binary had the new version running faster, but I assume that's just noise. I think observing a difference here would take a microbenchmark, but even then it shouldn't have an appreciable impact on real-world performance.